### DISCUSSION TITLE
Changing shape color of an existing PDF

### DISCUSSION THREAD

**Discussion by sumanmajumder123** [pymupdf/PyMuPDF#4002] [Looking for help] [answered]
Hello, 

First of all, I'm quite new in python, so my question is probably easy to answer, but I've tried somethings and didn't figure out how to solve my self.

I generate many PDF files with more than 300 pages. I already use a python code to reorder the pages and I want to recolor all the shapes of one certain color. I know I can do this with Illustrator, but the code will help me a lot.

Below is the code I'm using. Using debug I can see that "fill" dict is updated everytime, but once I open the new document the color of the shape is not the new one. I tried shape.update(), shape.commit() and other things.

```
import pymupdf 

def alterar_cor(input_pdf, output_pdf, cor_original, nova_cor):
    doc = pymupdf.open(input_pdf)
    for page in doc:
        for shape in page.get_drawings():
            
            if "fill" in shape and shape["fill"] == cor_original:
                shape["fill"] = nova_cor                  
            shape.update()

    doc.save(output_pdf)
    doc.close()

cor_original = (0.13669031858444214, 0.12195010483264923, 0.1252918243408203) 
nova_cor = (0.5, 0, 0) 
input_pdf = "C:/Users/Sum4n/Desktop/SMP/Gerar arquivos/Entrada.pdf"
output_pdf = "C:/Users/Sum4n/Desktop/SMP/Gerar arquivos/modificado.pdf"

alterar_cor(input_pdf, output_pdf, cor_original, nova_cor)
```
Thanks and sorry for my english.

**Comment by AumGupta** [comment]
Hey @sumanmajumder123,
Try calling `page.clean()` to force a redraw of the page content.

```python
import pymupdf 

def alterar_cor(input_pdf, output_pdf, cor_original, nova_cor):
	doc = pymupdf.open(input_pdf)
	for page in doc:
		for shape in page.get_drawings():
			if "fill" in shape and shape["fill"] == cor_original:
				shape["fill"] = nova_cor
				shape.update()
		page.clean()  # Force redraw of the page
	doc.save(output_pdf)
	doc.close()

cor_original = (0.13669031858444214, 0.12195010483264923, 0.1252918243408203) 
nova_cor = (0.5, 0, 0) 
input_pdf = "C:/Users/Sum4n/Desktop/SMP/Gerar arquivos/Entrada.pdf"
output_pdf = "C:/Users/Sum4n/Desktop/SMP/Gerar arquivos/modificado.pdf"

alterar_cor(input_pdf, output_pdf, cor_original, nova_cor)
```

**Comment by sumanmajumder123** [comment]
page.clean() returned this error: AttributeError: 'Page' object has no attribute 'clean'
page.clean_contents() doesn't change the results

**Comment by JorjMcKie** [comment, marked as answer]
There is no way of changing colors of existing objects on a PDF page!

**Comment by sumanmajumder123** [comment]
Ok. Thanks for the reply! Go back to Illustrator 😭

**Reply by JorjMcKie** [reply]
Before cutting edges:
When you create objects (images, text, shapes, ...) you can of course select colors you like.

Only if you are confronted with a PDF page made by someone / something else, then you cannot manipulate / edit it to just change colors and hope every thing else will remain untouched. **_That_** is not supported.

**Reply by sumanmajumder123** [reply]
Got it! Yep, I created new objects to test if I was doing something wrong with the code, the new ones worked like a charm! Thanks again
